### PR TITLE
Added deepChildren support from ArcEglos' pull request

### PR DESCRIPTION
**Pull request by ljcrapo**
**What kind of change does this PR introduce?**

Feature

**Did you add tests for your changes?**

Yes

**If relevant, link to documentation update:**

https://github.com/webpack/webpack.js.org/pull/1625

**Summary**

If you use the CommonsChunkPlugin with the children flag enabled it currently only cares about direct children. This change makes it so it cares about all children that are not required from somewhere else than the common chunk.
https://github.com/webpack/webpack/issues/3981

**Does this PR introduce a breaking change?**

If the config deepChildren isn't set the behavior stays the same as before, it only changes if the flag gets set.

**Other information**

Thanks to Florian Scholz (ArcEglos) for doing all the work.
https://github.com/webpack/webpack/pull/4255

### CODE REVIEW

**Comment by ljcrapo** [comment]
I see that the new test is failing in node v4. I'm working to resolve the issue.

**Comment by wbern** [comment]
Sweet! I am so waiting for this feature.

File: lib/optimize/CommonsChunkPlugin.js [hunk 1]
```diff
@@ -223,22 +257,9 @@ Take a look at the "name"/"names" or async/children option.`);
...
-				if(chunk.isInitial())
-					return false;
-
-				// we can only move modules from this chunk if the "commonChunk" is the only parent
-				if(!asyncOption)
-					return chunk.parents.length === 1;
-
-				return true;
-			});
+			let affectedChunks = [];
```
Comment (sokra) [inline lib/optimize/CommonsChunkPlugin.js]: Make it a `Set`

File: examples/common-chunk-grandchildren/README.md [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+
+});
+```
+
+# Info
+
+## Uncompressed
+
+```
+Hash: [1mfd4796594a8274a0759b[39m[22m
```
Comment (sokra) [inline examples/common-chunk-grandchildren/README.md]: This should not include the color special chars. Try to pass `--no-colors` in examples/build-common.js to fix it.

File: examples/common-chunk-grandchildren/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+	},
+	plugins: [
+		new webpack.optimize.CommonsChunkPlugin({
+			name: "main",
+			minChunks: 2,
+			children: true,
+			deepChildren: true,
+		})
+	]
+};
```
Comment (sokra) [inline examples/common-chunk-grandchildren/webpack.config.js]: Could you also a `async` mode to this example? You can pass multiple configurations to include both modes in a single example.

File: test/configCases/commons-chunk-plugin/move-to-grandparent/index.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+it("should correctly include indirect children in common chunk", function(done) {
+	Promise.all([
+		import('./pageA'),
+		import('./pageB')
+	]).then((imports) => {
+		imports[0].default.should.be.eql("reuse");
+		imports[1].default.should.be.eql("reuse");
+		done();
+	}).catch(e => {
+		done();
```
Comment (sokra) [inline test/configCases/commons-chunk-plugin/move-to-grandparent/index.js]: Pass the error to the done callback, elsewise errors are ignored.

File: examples/common-chunk-grandchildren/build.js [hunk 5]
```diff
@@ -0,0 +1 @@
+require("../build-common");
```
Comment (sokra) [inline examples/common-chunk-grandchildren/build.js]: add `global.NO_TARGET_ARGS = true;` before this line to disable the default and allow multiple configurations.

**Comment by ljcrapo** [comment]
@sokra or @timse what more needs to be done to get this merged in?

**Comment by ljcrapo** [comment]
I added a pull request to the documentation for this feature.

**Comment by sokra** [comment]
Thanks

**Comment by ZuBB** [comment]
@sokra are you going to cook one more minor release before webpack 4?

**Comment by wbern** [comment]
Can someone please look at #4392, specifically my question.

https://github.com/webpack/webpack/issues/4392#issuecomment-338153563

**Pull request merged**